include/mesh/mesh_base.h

```diff
@@ -263,6 +263,29 @@ class MeshBase : public ParallelObject
    */
   void set_elem_dimensions(const std::set<unsigned char> & elem_dims);
 
+  /**
+   * Tabulate a user-defined "code" for elements which belong to the element sets
+   * specified in \p id_set. Also sets up the inverse mapping, so that if one knows
+   * all the element sets an Elem belongs to, one can look up the corresponding code.
+   */
+  void add_elemset_code(dof_id_type code, const std::set<elemset_id_type> & id_set);
+
+  /**
+   * Determines the number of unique elemset ids which have been added
+   * via add_elemset_code() calls by looping over the
+   * _elemset_codes_inverse_map and counting them.
```

Comment:
That's kind of expensive for what's usually an O(1) type operation, if it might be called repeatedly.  Should we just keep a full `set<elemset_id_type>` in sync?  Or would doing that backfire by killing the asymptotic complexity of a future `remove_elemset_code`?

Reply:
Yeah, I was imagining that there would only be a few elemsets defined, but your comment makes sense in the case where somebody inevitably wants to define thousands of elemsets... I would be OK with adding an additional `set<elemset_id_type>` member to MeshBase, but I have not yet considered the consequences of a `remove_elemset()` capability. This would trigger a change to all the elemset codes involving that elemset... it seems messy.

Reply:
Oh, wow, extremely messy; I hadn't considered that.  On the other hand, it means we don't really have to care about how keeping a full `set<elemset_id_type>` would impact it - the difference between O(N_elem+1) and O(N_elem+N_elemsets) should be negligible.

Reply:
I added a new MeshBase member, `_all_elemset_ids` to support O(1) behavior of `MeshBase::n_elemsets()`.